### PR TITLE
backend-app-api, backend-test-utils: added backend.stop() and auto stop in tests

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

No more backends running all over the place :grin:

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by freben — COMMENTED** [review]
Neat.

File: packages/backend-test-utils/api-report.md [hunk 1]
```diff
@@ -23,13 +24,15 @@ export function setupRequestMockHandlers(worker: {
...
->(options: TestBackendOptions<TServices, TExtensionPoints>): Promise<void>;
+>(options: TestBackendOptions<TServices, TExtensionPoints>): Promise<Backend>;
 
 // @alpha (undocumented)
 export interface TestBackendOptions<
   TServices extends any[],
   TExtensionPoints extends any[],
 > {
+  // (undocumented)
+  autoStop?: boolean;
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-test-utils/api-report.md:35]: Yep, could convert this to `'afterEach' | 'never'` for now

[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-test-utils/api-report.md:35]: Alright. I just felt it's pretty common to fire up one instance of a thing and doing several tests against it. Whether it's a good idea or not :)

[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-test-utils/api-report.md:35]: Ah, I don't think we can do `'afterAll'` tbh, since it will always be scoped to the entire test file. I'm pretty sure that would lead to confusion if you're using it within a specific `describe` block that you expect it to be scoped to.

[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-test-utils/api-report.md:35]: Will it? You can trigger the afterAll within a describe and it'll apply only within that describe block eh? (Ie after all THOSE tests) Maybe I'm misinformed.

[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-test-utils/api-report.md]: Yep confirmed, that's what happens; `before*` and `after*` are "scoped".

File: .changeset/green-moose-leave.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/backend-test-utils': patch
+---
+
+Backends started with `startTestBackend` are now automatically stopped after the each test, unless the `autoStop` option is set to `false`.
```
Comment (freben) [inline .changeset/green-moose-leave.md:5]: Actually `'never'` below, but tbh `false` probably feels better 🤔  Unless one wants to stay consistently with strings

File: packages/backend-app-api/src/wiring/BackendInitializer.ts [hunk 3]
```diff
@@ -165,4 +167,24 @@ export class BackendInitializer {
 
     return orderedRegisterInits;
   }
+
+  async stop(): Promise<void> {
+    if (!this.#started) {
+      return;
+    }
+    this.#started = false;
```
Comment (freben) [inline packages/backend-app-api/src/wiring/BackendInitializer.ts:175]: Overly cautious perhaps but could even refuse to re-`start` after. Like a `state: 'idle' | 'started' | 'stopped'`. Or how about thoroughly `state: 'idle' | 'starting' | 'started' | 'stopping' | 'stopped'` capturing the async nature of the operations

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts [hunk 4]
```diff
@@ -101,4 +127,30 @@ export async function startTestBackend<
...
+  if (typeof afterEach !== 'function') {
+    return;
+  }
+  if (registered) {
+    return;
+  }
+  registered = true;
+
+  afterEach(async () => {
+    for (const backend of backendInstances) {
```
Comment (freben) [inline packages/backend-test-utils/src/next/wiring/TestBackend.ts:145]: `Promise.all`? 😄 Not that they're expected to take long, but every little bit matters i guess.

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts [hunk 5]
```diff
@@ -101,4 +127,30 @@ export async function startTestBackend<
...
+        await backend.stop();
+      } catch (error) {
+        console.error(`Failed to stop backend after test, ${error}`);
+      }
+    }
+    backendInstances.length = 0;
+  });
+}
+
+registerTestHooks();
```
Comment (freben) [inline packages/backend-test-utils/src/next/wiring/TestBackend.ts:156]: Hm. Given the scoping behavior of `before*` and `after*`, I wonder if this will actually not always work the way you intend it to. Feels like the only way of making things safe, is to **call** those functions at the exact time that the creation of the backend happens. I.e. you have to put their invocation directly inside `startTestBackend`. But the problem then, is that you can't do that inside an `it` - and it will be equally possible that `startTestBackend` is invoked at the root, or in a `describe`, or inside an `it`. All should ideally be made to work equally well, right? You should rewrite the test to exercise all of those cases and making sure they work and don't miss stopping or stop more than once.

Comment (freben) [reply packages/backend-test-utils/src/next/wiring/TestBackend.ts:156]: Hmm I should revise that perhaps ... I guess that what you have here works in practice 🤔 At least for afterEach, and as long as you do the creation inside an `it`. But now that the default value is `afterEach`, it'll mess up instances that were created in a `describe` if there is more than one `it`. Just highlighting that that's a pretty breaking behavior that might warrant upping the changeset level and have a special mention.

Comment (Rugvip) [reply packages/backend-test-utils/src/next/wiring/TestBackend.ts:156]: To sum up discussion on Discord: we'll try to implement automatic `afterAll` behavior as well, so that things behave correctly when starting a backend in a describe block. Probably using `expect.getState().currentTestName`

File: packages/backend-app-api/src/wiring/BackendInitializer.ts (hunk 3 above)
Comment (Rugvip) [reply packages/backend-app-api/src/wiring/BackendInitializer.ts:175]: Switched to not allowing re-starts at all.

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts (hunk 5 above)
Comment (Rugvip) [reply packages/backend-test-utils/src/next/wiring/TestBackend.ts:152]: Turns out `expect.getState().currentTestName` doesn't work :(, it doesn't reset state after tests

We settled for stopping at the end of all tests instead.

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts [hunk 6]
```diff
@@ -101,4 +123,32 @@ export async function startTestBackend<
   }
 
   await backend.start();
+
+  return backend;
 }
+
+let registered = false;
+function registerTestHooks() {
+  if (typeof afterEach !== 'function') {
```
Comment (freben) [inline packages/backend-test-utils/src/next/wiring/TestBackend.ts:132]: It's a bit funny that we check afterEach but call afterAll :) It'll work in practice, buuuut

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts [hunk 7]
```diff
@@ -47,11 +53,19 @@ export interface TestBackendOptions<
   features?: BackendFeature[];
 }
 
+const defaultServiceFactories = [
+  rootLoggerFactory(),
+  loggerFactory(),
+  lifecycleFactory(),
```
Comment (freben) [inline packages/backend-test-utils/src/next/wiring/TestBackend.ts:59]: I'd totally be up for a config service with an empty config, and perhaps a database service (that creates a sqlite database automagically if there's no backend.database key), ... and maybe some more.

Could be in a separate PR of course, if at all. It's a bit of a promise to make of course, to have them ever-present with a certain behaviour.

**Review by freben — APPROVED** [review]

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts (hunk 7 above)
Comment (Rugvip) [reply packages/backend-test-utils/src/next/wiring/TestBackend.ts:59]: Oh what, didn't realize these changes where part of this 😅. Either way, it's covered in #13082, yep separate PR

File: packages/backend-test-utils/src/next/wiring/TestBackend.ts (hunk 6 above)
Comment (Rugvip) [reply packages/backend-test-utils/src/next/wiring/TestBackend.ts:132]: doh
